frontend/src/views/admin/matching/index.tsx

```diff
@@ -21,11 +83,218 @@ export function AdminMatchingView() {
         }
     }, [activeSession, dispatch]);
 
+    React.useEffect(() => {
+        async function initializeMatches() {
+            const initialMatches: Match[] = [];
+
+            for (const applicant of applicants) {
+                const mostRecentApplication = GetNewestApplication(
+                    applicant,
+                    applications
+                );
+                if (!mostRecentApplication) {
+                    continue;
+                }
+
+                // Mark positions as being applied for
+                for (const positionPreference of mostRecentApplication.position_preferences) {
+                    initialMatches.push({
+                        applicantId: applicant.id,
+                        utorid: applicant.utorid,
+                        positionId: positionPreference.position.id,
+                        positionCode: positionPreference.position.position_code,
+                        status: "applied",
+                        hoursAssigned: 0,
+                    });
+                }
+            }
+
+            // Mark positions as being assigned
+            for (const assignment of assignments) {
+                const matchingAssignment = initialMatches.find(
+                    (match) =>
+                        match.applicantId === assignment.applicant.id &&
+                        match.positionId === assignment.position.id &&
+                        match.status === "applied"
+                );
+
+                // Update existing match object if it exists
+                if (matchingAssignment) {
+                    matchingAssignment.status = "assigned";
+                    matchingAssignment.hoursAssigned = assignment.hours
+                        ? assignment.hours
+                        : 0;
+                } else {
+                    // Otherwise, create a new one
+                    initialMatches.push({
+                        applicantId: assignment.applicant.id,
+                        utorid: assignment.applicant.utorid,
+                        positionId: assignment.position.id,
+                        positionCode: assignment.position.position_code,
+                        status: "assigned",
+                        hoursAssigned: assignment.hours ? assignment.hours : 0,
+                    });
+                }
+            }
+
+            return await dispatch(batchUpsertMatches(initialMatches));
+        }
+
+        initializeMatches();
+    }, [dispatch, applicants, assignments, applications]);
+
+    // Get information about positions
+    const positionSummaries = React.useMemo(() => {
```

Comment:
This is way too big to go in a component. Break some of this up into subfunctions. (they should probably be defined in a separate file. As a general rule, I like a single file containing no more than four "scrolls" worth of code.